crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -0,0 +1,1468 @@
+    fn can_blobs_fit_into_matrix(
+        blobs_sizes: &[usize],
+        n_columns: usize,
+        column_size: usize,
+    ) -> bool {
+        let mut used_cols = 0;
+        for &blob in blobs_sizes {
+            let cur = blob.div_ceil(column_size);
+            if used_cols + cur > n_columns {
+                return false;
+            }
+            used_cols += cur;
+        }
+
+        true
+    }
+
+    /// Get the data size of the quilt index.
+    pub fn get_quilt_index_data_size(combined_data: &[u8]) -> Result<usize, QuiltError> {
+        if combined_data.len() < QUILT_INDEX_SIZE_PREFIX_SIZE + QUILT_TYPE_SIZE {
+            return Err(QuiltError::FailedToExtractQuiltIndexSize);
+        }
+
+        let data_size = u64::from_le_bytes(
+            combined_data[QUILT_TYPE_SIZE..QUILT_INDEX_SIZE_PREFIX_SIZE + QUILT_TYPE_SIZE]
+                .try_into()
+                .map_err(|_| QuiltError::FailedToExtractQuiltIndexSize)?,
+        );
+        let data_size = usize::try_from(data_size).expect("data_size should fit in usize");
+
+        Ok(data_size)
+    }
+
+    pub fn get_quilt_version_byte(data: &[u8]) -> Result<u8, QuiltError> {
+        if data.is_empty() {
+            return Err(QuiltError::Other(
+                "failed to get quilt version byte: data is empty".to_string(),
+            ));
+        }
+        Ok(data[0])
+    }
+
+    /// Gets the ith column of data, as if `data` is a 2D matrix.
+    ///
+    /// # Arguments
+    /// * `i` - The column index.
+    /// * `data` - The data to extract the column from.
+    /// * `row_size` - The size of each row in bytes.
+    /// * `symbol_size` - The size of each symbol in bytes.
+    ///
+    /// # Returns
+    /// A vector containing the bytes from the ith column.
+    fn get_column(
+        i: usize,
+        data: &[u8],
+        row_size: usize,
+        symbol_size: usize,
+    ) -> Result<Vec<u8>, QuiltError> {
+        // Verify inputs make sense.
+        if row_size == 0
+            || data.is_empty()
+            || symbol_size == 0
+            || row_size % symbol_size != 0
+            || data.len() % row_size != 0
+        {
+            return Err(QuiltError::InvalidFormatNotAligned(format!(
+                "row_size: {}, symbol_size: {}, data.len(): {}",
+                row_size,
+                symbol_size,
+                data.len()
+            )));
+        }
+
+        let n_rows = data.len() / row_size;
+        if i >= (row_size / symbol_size) {
+            return Err(QuiltError::IndexOutOfBounds(i, row_size / symbol_size));
+        }
+
+        let mut column = Vec::with_capacity(n_rows * symbol_size);
+
+        for row in 0..n_rows {
+            let start_idx = row * row_size + i * symbol_size;
+            let end_idx = start_idx + symbol_size;
+
+            // Check if we have enough data for this chunk.
+            if end_idx > data.len() {
+                break;
+            }
```

Comment:
Question: Can this actually happen? AFAICT, the maximum value of `end_idx` is `(n_rows-1) * row_size + (row_size-symbol_size) + symbol_size = n_rows*row_size = data.len()`.